# Aula 2: Sistemas e multiplicação de matrizes

## Sistema, solução e tipos de sistemas

Um <b>sistema linear</b> com $m$ linhas e $n$ colunas é uma expressão do tipo:
$$\left\{\begin{array}{ccccccc}
a_{11}X_1 & + & \cdots & + & a_{1n}X_n & = & Y_1\\
\vdots     &     &          &     &      \vdots &     & \vdots  \\
a_{m1}X_1 & + & \cdots & + & a_{mn}X_n & = & Y_m\\
\end{array}\right.$$

com $a_{ij}, Y_{i} \in {\mathbb{R}}$. 

Uma $n$-upla $\mathbf{x} = (x_1,\dots, x_n) \in {\mathbb{R}}^{n}$ é chamada de <b>solução</b> do sistema se valem simultaneamente as seguintes equações
$$\begin{array}{ccccccc}
a_{11}x_1 & + & \cdots & + & a_{1n}x_n & = & Y_1\\
\vdots     &     &          &     &      \vdots &     & \vdots  \\
a_{m1}x_1 & + & \cdots & + & a_{mn}x_n & = & Y_m\\
\end{array}$$

Se $Y_1 = Y_2 = \cdots = Y_m=0$, então dizemos que o sistema é <b>homogêneo</b>.

Se o sistema possuir uma única solução, dizemos que o sistema é <b>possível e determinado</b>.

Se o sistema possuir mais de uma solução, dizemos que o sistema é <b>possível e indeterminado</b>.

Se o sistema não possuir solução, dizemos que o sistema é <b>impossível</b>.

Diremos que dois sistemas lineares são <b>equivalentes</b> se cada equação de um sistema for combinação linear formal das equações do outro.

<b>Teorema</b> Sistemas lineares equivalentes possuem o mesmo conjunto de soluções.

<b>Observação:</b> Se sistemas equivalentes possuem o mesmo conjunto soluções é inteligente buscarmos o sistema equivalente mais simples de ser resolvido. Qual seria esse sistema? Nas próximas aulas buscaremos a resposta para essa pergunta.

## Exercício teórico

Determine o conjunto $S$ formado por todas as soluções do sistema linear homogêneo dado por
$$\left\{\begin{array}{ccccc}
x & - & y & = & 0\\
2x & + & y & = & 0
\end{array}\right.$$

<b>Resposta:</b> $S = \{ (0,0) \}$

## Exemplo prático

Numa sorveteria Matias pagou R\\$ 29,00 por 3 bolas de sorvete de morango e 5 bolas de sorvete de chocolate. Já Fred pagou R\\$ 31,00 por 5 bolas de sorvete de morango e 4 bolas de sorvete de chocolate nesta mesma sorveteria. Quanto custou cada bola de sorvete de morando e cada bola de sorte de chocolate?

<b>Resposta:</b>

Sendo x o preço da bola de sorvete de morando e y o preço da bola de sorvete de chocolate temos que
$$\left\{\begin{array}{ccccc}
3x & + & 5y & = & 29\\
5x & + & 4y & = & 31
\end{array}\right.$$
E, portanto, $x=3$ e $y=4$.

## Sistemas lineares com Python

Para resolver o sistema do exemplo prático acima utilizaremos a função <i>solve</i> da biblioteca scipy. Para isso devemos importá-la:

In [7]:
from scipy.linalg import solve

Nosso próximo passo é definir uma matriz quadrada $A = (a_{ij})$ com os coeficientes do sistema do exemplo anterior:

In [10]:
import numpy as np

A = np.array([
    [3, 5],
    [5, 4]
])

Precisaremos também de uma matriz com uma coluna $Y = (Y_1\ Y_2)^T$ com os valores pagos por cada pessoa:

In [11]:
Y = np.array([
    [29],
    [31]
])

Agora basta passar como parâmetros da função <i>solve</i> as matrizes $A$ e $Y$:

In [12]:
solve(A, Y)

array([[3.],
       [4.]])

<b>Dica:</b> Novamente, aqui temos um exemplo simples para entender conceitos poderosos. O primeiro é a representação matemática do problema no contexto de sistemas lineares. O segundo é o uso de uma ferramenta para resolver o problema.

## Multiplicação de matrizes

Sejam $A = (a_{ij}) \in {\cal{M}}_{m\times n}$ e $B = (b_{ij}) \in {\cal{M}}_{n\times p}$. Definimos a <b>multiplicação (ou produto) 
de $A$ com $B$</b>, denotado por $A\ast B$, ou por $AB$, como sendo a matriz $C=(c_{ij})\in {\cal{M}}_{m\times p}$, definida por
$$C = (c_{ij}) = \left( \sum^{n}_{k=1}a_{ik}b_{kj} \right).$$

## Exemplo teórico

Sejam 
$$A = \left( \begin{array}{ccc} 2 & 1 & 0 \\ -1 & 3 & 2 \end{array} \right)\ \ \text{e} \ \ B = \left( \begin{array}{cc} 1 & 2 \\ 2 & 1 \\ 0 & 0 \end{array}\right)$$
matrizes reais. Determine $C = AB$.

<b>Resposta:</b>
$$ C = AB = \left( \begin{array}{cc} 4 & 5 \\ 5 & 1 \end{array} \right).$$

## Multiplicação de matrizes com Python

Para multiplicar duas matrizes em Python podemos utilizar a função <i>matmul</i> da biblioteca numpy:

In [14]:
# definindo A
A = np.array([
    [2,1,0],
    [-1,3,2]
])
# definindo B
B = np.array([
    [1,2],
    [2,1],
    [0,0]
])
# calculando C = AB
C = np.matmul(A,B)
# mostrando a matriz C
print(C)

[[4 5]
 [5 1]]


## Exemplo prático de multiplicação de matrizes

Vamos considerar ainda a sorveteria do exemplo prático anterior. Sabendo que além do Matias e do Fred, também compraram bolas de sorvete de morango e de chocolate Bia, Bob, Ted, Ana, João, Elias e Raquel. Veja a tabela a seguir:

<center><img src="assets/sorvetes.png" alt="sorvetes" width="200px"></center>

Quanto a sorveteria obteve com as venda desses sorvetes?

<b>Resposta</b>

Podemos fazer na mão essas contas, mas pense numa empresa grande com muitos clientes e muitos produtos, isso não seria viável. Vamos representar a planilha acima através de uma matriz:

In [16]:
A = np.array([
    [3,5],
    [5,4],
    [2,4],
    [5,3],
    [3,2],
    [5,1],
    [1,2],
    [4,5],
    [2,7]
])

Agora vamos colocar o preço de cada bola numa matriz com uma coluna:

In [18]:
B = np.array([
    [3],
    [4]
])

Observe que cada linha da matriz $C = AB$ representa quanto cada pessoa gastou. Pare um tempo para entender essa observação.

In [20]:
C = np.matmul(A, B)
print(C)

[[29]
 [31]
 [22]
 [27]
 [17]
 [19]
 [11]
 [32]
 [34]]


Vamos então somar tudo:

In [21]:
sum(C)

array([222])

Logo, a sorveteria obteve 222 reais com essa venda!

## Exemplo prático de multiplicação de matrizes

Na pasta datasets tem uma planilha do tipo .csv com a mesma matriz do exemplo anterior. Gostariamos de extrair os dados diretamente desta planilha para evitar digitar na mão todas as vendas.

Para isso utilizaremos a função <i>read_csv</i> da biblioteca pandas.

In [28]:
import pandas as pd

planilha = pd.read_csv('datasets/sorvetes.csv', sep=";", index_col=0)

Vejamos como fica a variável planilha:

In [29]:
planilha

Morango  Chocolate
Matias        3          5
Fred          5          4
Bia           2          4
Bob           5          3
Ted           3          2
Ana           5          1
João          1          2
Elias         4          5
Raquel        2          7

Para colocar os valores da planilha numa matriz como fizemos no exemplo anterior:

In [34]:
A = planilha.values
A

array([[3, 5],
       [5, 4],
       [2, 4],
       [5, 3],
       [3, 2],
       [5, 1],
       [1, 2],
       [4, 5],
       [2, 7]], dtype=int64)

Podemos então prosseguir como no exemplo anterior:

In [35]:
B = np.array([
    [3],
    [4]
])
C = np.matmul(A, B)
sum(C)

array([222], dtype=int64)

## Exercícios

<b>Exercício 1</b> Numa sorveteria Matias pagou R\\$ 44,00 por 3 bolas de sorvete de morango, 5 bolas de sorvete de chocolate e 2 bolas de sorvete de pistache. Já Fred pagou R\\$ 28,00 por 1 bola de sorvete de morango, 3 bolas de sorvete de chocolate e 2 bolas de sorvete de pistache. Por fim, Ana pagou R\\$ 31,00 por 2 bolas de sorvete de morango, 4 bolas de sorvete de chocolate e 1 bola de sorvete de pistache nesta mesma sorveteria. Quanto custou cada bola de sorvete de morango, cada bola de sorte de chocolate e cada bola de sorvete de pistache?

<b>Exercício 2</b> Considerando a mesma sorveteria do Exercício 1 e o arquivo <i>sorvetes_exercicio.csv</i> dentro da pasta datasets que representa as vendas de um único dia nesta sorveteria, calcule quanto a sorveteria obteve com as vendas.